### PR TITLE
Reading Realm values from a `Realm.Dictionary`

### PR DESCRIPTION
**Pull request by kraenhansen**
I believe this closer resembles the access pattern used by the MMKV test function, as this will access values from a dictionary of arbitrary key/value pairs. There is no pre-fetching or caching of the values in the Realm SDK layer and the database is hit on every get as with MMKV (there are of course some at the lower levels, as with any database).

These are the results I'm getting on my M1 in an iOS simulator.

```
 LOG  Running Benchmark in 3... 2... 1...
 LOG  Starting Benchmark "MMKV                 "...
 LOG  Finished Benchmark "MMKV                 "! Took 10.8298ms!
 LOG  Starting Benchmark "MMKV Encrypt         "...
 LOG  Finished Benchmark "MMKV Encrypt         "! Took 8.3246ms!
 LOG  Starting Benchmark "AsyncStorage         "...
 LOG  Finished Benchmark "AsyncStorage         "! Took 188.2416ms!
 LOG  Starting Benchmark "Expo Secure Storage  "...
 LOG  Finished Benchmark "Expo Secure Storage  "! Took 980.0811ms!
 LOG  Starting Benchmark "React Native Keychain"...
 LOG  Finished Benchmark "React Native Keychain"! Took 566.4805ms!
 LOG  Starting Benchmark "SQLite               "...
 LOG  Finished Benchmark "SQLite               "! Took 45.7815ms!
 LOG  Starting Benchmark "RealmDB              "...
 LOG  Finished Benchmark "RealmDB              "! Took 11.4357ms!
 LOG  Starting Benchmark "WatermelonDB         "...
 LOG  Finished Benchmark "WatermelonDB         "! Took 35.5760ms!
 ```

### CODE REVIEW

**Comment by mrousavy** [comment]
Hey thanks for your PR. As I just replied in the previous PR, I don't think this is realistic. These are smart tricks to make Realm appear faster in a Benchmark, which should be a close to real world scenario of how users might use your database.

In no real world production app are users going to fully pre-cache every single query like this using dictionaries. People are going to use getters, and I think that's what this Benchmark should reflect.

**Comment by kraenhansen** [comment]
Please help me understand how this is different from the access pattern that the MMKV benchmark does? The `object` that I'm pulling from the database is a "live object", not a plain data object, the same with `dict`. And accessing a key from it via a string index accessor is just syntactic sugar on top of the underlying getter that is indeed hitting the database.

> should be a close to real world scenario of how users might use your database.

This is a totally valid way to access the database, if you want to use it as a fast KV store.
That being said, these benchmarks are nowhere close to resembling real life access patterns 😊 (as they're getting the same value in a loop, with no updates or deletions interleaved). They're highly synthetic, which is fine in my opinion, but to be fair they should at least compare similar operations?

I'd love to have a higher bandwidth conversation to discuss this in detail 🙂

**Comment by mrousavy** [comment]
> Please help me understand how this is different from the access pattern that the MMKV benchmark does?

Because the `getValueFromRealm()` function uses a cached/pre-warmed object that is hot on a specific key already. This is an optimization that you do not do in a real world app, instead you access values through their key. Which is what this does not do. `getValueFromRealm` does not pass in a key.

I understand this might just be an API design difference, but if you rely on keeping the "get query" (in this case: "get `Test`") warmed up, then it just isn't a good comparison.
MMKV passes the key in `getValueFromMMKV` in each time.

**Comment by kraenhansen** [comment]
Just to clarify and make this painfully obvious: We use [a Proxy to trap the property get](https://github.com/realm/realm-js/blob/05deac9b58143f654698d06f8766a840ae9af2c8/packages/realm/src/Dictionary.ts#L48) (when getting `k` off `dict`) and then pass the key to the database when getting the value. As such, I don't agree that it's "warmed up", you could get / set any value on `dict` (setting would require a write transaction) and it would persist and read from the database on any of these operations.

**Comment by mrousavy** [comment]
But you're not passing a new key in each time, this query only happens once, right?

File: app/src/storages/Realm.ts [hunk 1]
```diff
@@ -3,25 +3,26 @@ import Realm from 'realm';
...
-  });
+  return realm.create<Test>('Test', {dict: {k: 'hello'}});
 });
 
+const dict = object.dict;
+
 export function getFromRealm() {
-  const object = realm.objectForPrimaryKey('Test', 1);
-  return object.value;
+  return dict.k;
```
Comment (kraenhansen) [inline app/src/storages/Realm.ts:27]: This is syntactic sugar for getting the value associated with the `"k"` key in the dictionary, it's not already read from the database, nor cached, as the `Realm.Dictionary` lazy-loading values from the database and getting this "k" hits a Proxy "get" trap:
- On `realm@11`: https://github.com/realm/realm-js/blob/v11.0.0-rc.0/lib/dictionary.js#L22-L48
- On `realm@12`: https://github.com/realm/realm-js/blob/main/packages/realm/src/Dictionary.ts#L46-L55

**Comment by kraenhansen** [comment]
> But you're not passing a new key in each time, this query only happens once, right?

We're not passing a _new_ key, no. It's always "k", just like the MMKV test use a const and always passes "k" on every get: https://github.com/mrousavy/StorageBenchmark/blob/3a0bb57081a06c11d49cd11c006b4ece529393ee/app/src/storages/MMKV.ts#L7

> this query only happens once, right?

The value is read from the database, not a JS-layer cache. The values isn't stored as a property on the `dict` object, but as I mentioned earlier, we're using a Proxy to trap the get and read the value from the database. We wouldn't technically call that a query (since there's no query string involved), but it does indeed read from the underlying database and if some other piece of code (possibly a separate thread or process working on the same database file) were mutating this value in the database, it would reflect.

While I understand the intention and limitations of this benchmark (it's supposed to just give a quick and dirty comparison), I do think it could be interesting if it was just slightly more involved to the point that it was easier to convince the reader of the test code, that it's actually hitting the database 👍

**Comment by pvinis** [comment]
addressed in https://github.com/pvinis/StorageBenchmark. if you think it's the wrong code, open up a similar PR to this, over there, and I'll merge 👍.

**Pull request closed without merging**